### DISCUSSION TITLE
about  curl_multi_poll

### DISCUSSION THREAD

**Discussion by xuquan316** [curl/curl#12541] [Q&A] [answered]
code:
~~~c
for(;;){
      ...
      curl_multi_perform(...);
      ...
      mc = curl_multi_poll(m_multi_handle, NULL, 0, 1000, &nums_fds);
      if(mc == CURLM_OK && nums_fds >0){
            while((handle =curl_multi_info_read(....) )!= null){
                int64_t current_time_recv = now;
                if (resp->msg == CURLMSG_DONE){
                         task_queue.push(resp->easy_handle);  //  find map <CURL*,void* data>,data used callback
                }
            }
      }
      .....
}

size_t rest_read_callback(....){
       ....
       int64_t current_time_callback = now;    //The callback is called once and the message is received.
       ....
}
~~~

Message request interval is 100ms
test:    current_time_recv -  current_time_callback （> 1ms and <200ms，40ms delay for most) ， Is there something I'm not using correctly?

init func:
    curl_multi_setopt(m_multi_handle, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);
    curl_multi_setopt(m_multi_handle, CURLMOPT_MAX_HOST_CONNECTIONS, m_connect_num); //1 or 10，http1.1 dont use one
    curl_multi_setopt(m_multi_handle, CURLMOPT_MAXCONNECTS, m_connect_num);

**Comment by xuquan316** [comment]
code:
~~~c
for(;;){
        request_queue.pop(data);
        curl_easy_deal...
        curl_multi_add_handle(.....);
        ...
        curl_multi_perform(...);
        ...
        mc = curl_multi_poll(m_multi_handle, NULL, 0, 1000, &nums_fds);
        if(mc == CURLM_OK && nums_fds >0){
                while((handle =curl_multi_info_read(....) )!= null){
                        int64_t current_time_recv = now;
                        if (resp->msg == CURLMSG_DONE){
                        task_queue.push(resp->easy_handle); // find map <CURL*,void* data>,data used callback
                        curl_multi_remove_handle(...);
                }
        }
    }
    .....
}

size_t rest_read_callback(....){
        ....
        int64_t current_time_callback = now; //The callback is called once and the message is received.
        ....
}
~~~
Message request interval is 100ms
test: current_time_recv - current_time_callback （> 1ms and <200ms，40ms delay for most) ， Is there something I'm not using correctly?

init func:
      curl_multi_setopt(m_multi_handle, CURLMOPT_PIPELINING, CURLPIPE_MULTIPLEX);
      curl_multi_setopt(m_multi_handle, CURLMOPT_MAX_HOST_CONNECTIONS, m_connect_num); //1 or 10，http1.1 dont use one
      curl_multi_setopt(m_multi_handle, CURLMOPT_MAXCONNECTS, m_connect_num);
another thread:
     request_queue.push(data);
     curl_multi_wakeup(m_multi_handle);

**Comment by jay** [comment]
> Message request interval is 100ms
> test: current_time_recv - current_time_callback （> 1ms and <200ms，40ms delay for most) ， Is there something I'm not using correctly?

I don't understand the question

**Comment by xuquan316** [comment]
> > Message request interval is 100ms
> > test: current_time_recv - current_time_callback （> 1ms and <200ms，40ms delay for most) ， Is there something I'm not using correctly?
> 
> I don't understand the question

Callback timing current_time_callback, and curl_multi_poll interface call timing current_time_recv, the time difference between the two 40ms

**Reply by bagder** [reply]
That's impossible to answer without a lot of more information.

**Reply by xuquan316** [reply]
After the callback function is called, wait about 40 ms before triggering the curl_multi_poll API call 

Shouldn't the event be triggered in real time?

**Reply by jay** [reply]
As far as I understand your question, you want to know why there's 40ms from the last time the read callback is called until the time the transfer ends. That's not an unusual amount of time. It could be any number of things but a chunk of it will certainly be the time it takes for the server response to be received.

**Reply by xuquan316** [reply]
compare，Tested 10,000 times【Calling callback once receives the message in its entirety】

use easy api
1 call  curl_easy_perform ==(20ms)==> (callback) recv response

use multi api
2 curl_multi_add_handle=>curl_multi_perform==(20ms)==>callback  ==(40ms)==>curl_multi_poll=>curl_multi_read...   

The question is:
（multi api）callback  ==(40ms)==>curl_multi_poll  ，Where does the 40ms come from here, shouldn't it be event real time?

**Reply by jay** [reply]
It depends on when you take the timestamps and what happens in between. Internally curl_easy_perform is a wrapper around multi, so there should not be a difference with a single transfer except that your multi logic may be different.

https://github.com/curl/curl/blob/7161cb17c01dcff1dc5bf89a18437d9d729f1ecd/lib/easy.c#L659-L771

**Reply by xuquan316** [reply]
Hello, can you help to write a use case where one thread sends a message and one thread is responsible for handling curl_multi_poll events（I feel like I'm using it a little bit incorrectly）
thanks

**Reply by bagder** [reply]
I would probably recommend you don't use such architecture/design: let the multi interface run in a single thread.

**Reply by jay** [reply]
Maybe [curl_multi_wakeup](https://curl.se/libcurl/c/curl_multi_wakeup.html) would help?

**Reply by xuquan316** [reply]
Every time "curl" is reinitialized, the "wakeup api" is called.  But the effect is the same.

**Reply by xuquan316** [reply]
Some of the data is provided by another thread, using a lockless queue, which uses "add perform poll remove" to the same effect.

**Reply by jay** [reply]
No information in this discussion indicates a bug in curl. Also, a lock free queue may have worse performance than locking.

**Comment by xuquan316** [comment, marked as answer]
> Maybe [curl_multi_wakeup](https://curl.se/libcurl/c/curl_multi_wakeup.html) would help?


Some of the data is provided by another thread, using a lockless queue, which uses "add perform poll remove" to the same effect.
> I would probably recommend you don't use such architecture/design: let the multi interface run in a single thread.
